pkg/pma/pma.c

```diff
@@ -0,0 +1,783 @@
+#include "pma.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "page.h"
+#include "sigsegv.h"
+#include "util.h"
+#include "wal.h"
+
+#ifdef PMA_TEST
+#    define inline_
+#    define static_
+#else
+#    define inline_ inline
+#    define static_ static
+#endif
+
+#define PMA_DEBUG
+
+//==============================================================================
+// CONSTANTS
+
+/// Number of bits in a byte.
+static_ const size_t kBitsPerByte = 8;
```

Comment:
Somewhat unrelated, but where does this style of "lower camel case" for *some* constants come from?

Reply:
All constants in `pkg/pma` are prefixed with a `k` and written in upper camel case to distinguish them from other identifiers; if you see an identifier starting with `k` in upper camel case, you know it's a constant and therefore will never change, reducing the amount of information you need to keep in your head (such as "where might this value get changed?").

Reply:
Why not `K_BITS_PER_BYTE`, in that case? Upper snake case is a widely used for constants and the `K_` can still represent that it's defined in `pkg/pma` .

Reply:
It's purely stylistic. We can bikeshed endlessly here. The point is to clearly communicate that a given symbol represents a constant value.

Reply:
I am strongly in favor, for what it's worth, of deciding on a single style and configuring `clang-format` and `clang-tidy` (or similar) to programmatically enforce that style. In lieu of that, which has met with resistance in the past, I'm simply using a style that I've found in my experience is most clear and straightforward in communicating the intentions of C code.

Reply:
@barter-simsum why are you against using `clang-format` or similar? Genuinely curious.

Reply:
I'm also curious why you're opposed to linters.

Reply:
@barter-simsum Also curious what's wrong with linters / auto-formatting.

@mcevoypeter re: bike-shedding - Sure, you should consider this conversation non-blocking for the purposes of merging. Just seems odd that given a professed preference for idiomatic C to not declare system-level constants as macros, and use standard macro naming conventions for them.

Reply:
If the linter was as unobtrusive as like the linux kernel's [checkpatch](https://docs.kernel.org/dev-tools/checkpatch.html) I wouldn't mind it much, but most are not. Note the disclaimer:

"Checkpatch is not always right. Your judgement takes precedence over checkpatch messages. If your code looks better with the violations, then its probably best left alone."

Highly opinionated linters really rose in popularity with javascript (sacred airbnb style etc). Prior to that, style guidelines were often just that, guidelines, not law. My experience with these types has been overwhelmingly negative. Code run through these things is often _not_ more readable, just mechanistically more consistent. I also dislike the comment litter they promote if you want to disable certain rules within a region of a file.

strict characters per line, where I should and shouldn't put a comma, opinions on variable names, and var alignment, braces always or braces never, [can't use 0xcafebabe](https://github.com/rust-lang/rust/commit/0d55bd11009686dba662b0f4e697183691e7b308) (lol) literals and similar, etc etc ad nauseam -- all of this is just hell and then we hook into into the build system.

I don't care if you use a linter to review your own code. They might help find conditions to collapse, unclear conditional expressions, etc, but I definitely don't want it enshrined in the build.

And no, I don't  think they reduce "bikeshedding." We just bikeshed about the linter rules instead of the code change itself.

Reply:
No codebase can meet the style preferences of everyone working on it if there are even two developers. Even if there's only one developer, it still won't meet the style preferences of everyone reading it. The only thing that you can hope for is consistency.

Having some auto-formatter or linter removes this factor from consideration; it removes an axis on which to bikeshed. You're right that there's still one ultimate bikeshed which is "what do we establish the rules to be?". However, it's a much easier problem with which to grapple once the system is in place: just make the top-level code owner for each lib supreme styling dictator.

> I don't care if you use a linter to review your own code... but I definitely don't want it enshrined in the build.

I think that we might be talking past each other here somewhat. I don't think anyone is pitching to connect a strict linter to the build workflow. What I'm hearing pitched is the inclusion of a `clang-format` config file in which the ultimate style rules for the component reside. This is to be used as a helper tool, so that the developer can write his code however he wants and run this tool once at the end before submitting PR. Maybe I'm the one who's misinterpreting what @mcevoypeter and @matthew-levan would like to have.

> My experience with these types has been overwhelmingly negative. Code run through these things is often not more readable, just mechanistically more consistent.

We're just going to have to agree to disagree here. I found this PR very readable.

> strict characters per line, where I should and shouldn't put a comma, opinions on variable names, and var alignment, braces always or braces never, can't use 0xcafebabe literals and similar, etc etc ad nauseam

We already have most of these things in `u3`, they're just manually enforced so contributors have to actively think about them instead of knowing that there's an auto-format tool that'll do it for them. Also, just because Rust chose stupid rules [doesn't mean we have to](https://github.com/ashelkovnykov/pma_malloc/blob/master/src/malloc.c#L146).

Reply:
> The only thing that you can hope for is consistency

Readability. Not consistency. the former is not context free. The latter is what linters enforce.

> I don't think anyone is pitching to connect a strict linter to the build workflow

This is exactly the case in new mars now (recently). I don't think intent to integrate linting into the bazel build of vere is at all an unreasonable assumption. https://github.com/urbit/new-mars/pull/39

> We're just going to have to agree to disagree here. I found this PR very readable

After some of the style changes Peter made, outside of a few extra long functions (made longer by 1 arg per line funcall formatting),  I have few complaints about the general style of this PR. The below style I find slightly hard to read, but I didn't comment on it, because I really don't care. I do not, however, want such a style autistically _enforced_ be it programmatic or informal.

```
            if (mmap(ptr,
                     kPageSz,
                     PROT_READ,
                     MAP_FIXED | MAP_PRIVATE,
                     fd,
                     offset_)
                == MAP_FAILED)
```

> We already have most of these things in `u3`

We have style _guidelines_ in u3, you're free to break them if reasonable. See some of the formatting of the pointer compression pr: https://github.com/urbit/vere/pull/164/files#diff-880188529bb675cec6511e9d295a25921cd0b0f95d7c7c4c14bb2c2dbbd2d6f7R2074.

> just make the top-level code owner for each lib supreme styling dictator

this encourages a proliferation of libs. We have too many Caesars already. There should just be one benevolent vere dicatator. No one claims this title afaik, could be @joemfb or @belisarius222. (Or a state/deepstate type deal like with Linus and gregkh)

---

My biggest overall complaint with any of these efforts, is they just get in the way of writing code.

To quote Ryan Dahl: "If you think it would be cute to align all of the equals signs in your code, if you spend time configuring your window manager or editor, if put unicode check marks in your test runner, if you add unnecessary hierarchies in your code directories, if you are doing anything beyond just solving the problem - you don't understand how fucked the whole thing is"

Please don't add any more tools that yell at me, especially for trivial syntactical reasons. Please, in code reviews, __avoid commenting on style__, except for truly odd choices (like the `_` suffix style which was addressed). Basic consistency is easily maintained. Who is harmed by slight deviations: a mul `*` unwrapped in spaces, a 1-line condition without swaddling braces, a switch case that allows fall through, etc?

I won't discuss this further in github comments. It's not relevant to this pr. We can settle it on the field with pistols. @joemfb is my second. Name yours ;)